designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,531 @@
+- Start Date: 2019-09-28
+- RFC PR: https://github.com/eslint/rfcs/pull/40
+- Authors: Toru Nagashima ([@mysticatea](https://github.com/mysticatea))
+
+# `ESLint` Class Replacing `CLIEngine`
+
+## Summary
+
+This RFC adds a new class `ESLint` that provides asynchronous API and deprecates `CLIEngine`.
+
+## Motivation
+
+- We have functionality that cannot be supported with the current synchronous API. For example, ESLint verifying files in parallel, formatters printing progress state, formatters printing results in streams etc. A move to an asynchronous API would be beneficial and a new `ESLint` class can be created with an async API in mind from the start.
+- Node.js will support [ES modules](https://nodejs.org/api/esm.html) stably on `13.0.0` at last. Node.js doesn't provide any way that loads ES modules synchronously from CJS. This means that ESLint (CJS) cannot load configs/plugins that are written as ES modules synchronously. Migrating to asynchronous API opens up doors to support those.
```

Comment:
@nzakas ESM will be unfledged in node 13 this month; it will happen very quickly.

Reply:
Ah good to know, thanks.

Reply:
Yes, it likely will happen `13.2.0` on 2019-11-19 because the unflagged has been merged (https://github.com/nodejs/node/pull/29866). I should update this sentence after the release.

Reply:
ES Modules syntax is now available.